database.go

```diff
@@ -0,0 +1,492 @@
+	task, err := t.GetBaseTask(id)
+	if err != nil {
+		return task, err
+	}
+
+	behaviors, err := t.GetTaskBehavior(id)
+	if err != nil {
+		return task, err
+	}
+	task.Behaviors = behaviors
+
+	statuses, err := t.GetTaskStatus(id)
+	if err != nil {
+		return task, err
+	}
+	task.Status = statuses
+
+	return task, err
+}
+
+// GetTaskBehavior fetches a task's set of behaviors from the DB by ID
+func (t *DBTx) GetTaskBehavior(id string) ([]AsyncTaskBehavior, error) {
+	query := `SELECT behavior_type, data FROM async_task_behavior WHERE async_task_id::text = $1`
+
+	rows, err := t.tx.Query(query, id)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var behaviors []AsyncTaskBehavior
+	for rows.Next() {
+		var dbbehavior DBTaskBehavior
+		if err := rows.Scan(&dbbehavior.BehaviorType, &dbbehavior.Data); err != nil {
+			return nil, err
+		}
+
+		behavior := AsyncTaskBehavior{BehaviorType: dbbehavior.BehaviorType}
+		if dbbehavior.Data.Valid {
+			jsonData := make(map[string]interface{})
+
+			err = json.Unmarshal([]byte(dbbehavior.Data.String), &jsonData)
+			if err != nil {
+				return behaviors, err
+			}
+
+			behavior.Data = jsonData
+		}
+
+		behaviors = append(behaviors, behavior)
+	}
+
+	return behaviors, nil
+}
+
+// GetTaskStatus fetches a tasks's list of statuses from the DB by ID, ordered by creation date
+func (t *DBTx) GetTaskStatus(id string) ([]AsyncTaskStatus, error) {
+	query := `SELECT status, created_date at time zone (select current_setting('TIMEZONE')) AS created_date
+	            FROM async_task_status WHERE async_task_id::text = $1 ORDER BY created_date ASC`
+
+	rows, err := t.tx.Query(query, id)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var statuses []AsyncTaskStatus
+	for rows.Next() {
+		var status AsyncTaskStatus
+		if err := rows.Scan(&status.Status, &status.CreatedDate); err != nil {
+			return nil, err
+		}
+
+		statuses = append(statuses, status)
+	}
+
+	return statuses, nil
+}
+
+type TaskFilter struct {
+	IDs             []string
+	Types           []string
+	Usernames       []string
+	StartDateSince  []time.Time
+	StartDateBefore []time.Time
+	EndDateSince    []time.Time
+	EndDateBefore   []time.Time
+	IncludeNullEnd  bool
+	Statuses        []string
+}
+
+// GetTasksByFilter fetches a set of tasks by a set of provided filters
+func (t *DBTx) GetTasksByFilter(filters TaskFilter) ([]AsyncTask, error) {
+	log.Info(filters)
+
+	var tasks []AsyncTask
+	var args []interface{}
+	var wheres []string
+
+	query := `SELECT id, type, username, data,
+	                 start_date at time zone (select current_setting('TIMEZONE')) AS start_date,
+			 end_date at time zone (select current_setting('TIMEZONE')) AS end_date
+	            FROM async_tasks`
+
+	currentIndex := 1
+
+	if len(filters.IDs) > 0 {
+		wheres = append(wheres, fmt.Sprintf(" id::text = ANY($%d)", currentIndex))
+		args = append(args, pq.Array(filters.IDs))
+		currentIndex = currentIndex + 1
+	}
```

Comment:
If someone's got a better idea for this sort of query building, I'm all ears. I didn't find any libraries I really liked, but I might not have dug enough.